sourmash_lib/lca/lca_utils.py

```diff
@@ -0,0 +1,302 @@
+    ancestor agreements/confusion.
+    """
+    if initial is None:
+        tree = {}
+    else:
+        tree = initial
+
+    if not assignments:
+        raise ValueError("empty assignment passed to build_tree")
+
+    for assignment in assignments:
+        node = tree
+
+        for lineage_tup in assignment:
+            if lineage_tup.name:
+                child = node.get(lineage_tup, {})
+                node[lineage_tup] = child
+
+                # shift -> down in tree
+                node = child
+
+    return tree
+
+
+def find_lca(tree):
+    """
+    Given a tree produced by 'find_tree', find the first node with multiple
+    children, OR the only leaf in the tree.  Return (lineage_tup, reason),
+    where 'reason' is the number of children of the returned node, i.e.e
+    0 if it's a leaf and > 1 if it's an internal node.
+    """
+
+    node = tree
+    lineage = []
+    while 1:
+        if len(node) == 1:                # descend to only child; track path
+            lineage_tup = next(iter(node.keys()))
+            lineage.append(lineage_tup)
+            node = node[lineage_tup]
+        elif len(node) == 0:              # at leaf; end
+            return tuple(lineage), 0
+        else:                             # len(node) > 1 => confusion!!
+            return tuple(lineage), len(node)
+
+
+class LCA_Database(object):
+    """
+    Wrapper class for taxonomic database.
+
+    obj.lineage_dict: key 'lineage_id' => lineage tuple [(name, rank), ...]
+    obj.hashval_to_lineage_id: key 'hashval' => 'lineage_id'
+    obj.ksize: k-mer size
+    obj.scaled: scaled value
+    obj.signatures_to_lineage: key 'md5sum' => 'lineage_id'
+    """
+    def __init__(self):
+        self.lineage_dict = None
+        self.hashval_to_lineage_id = None
+        self.ksize = None
+        self.scaled = None
+        self.signatures_to_lineage = None
+
+    def load(self, db_name):
+        "Load from a JSON file."
+        xopen = open
+        if db_name.endswith('.gz'):
+            xopen = gzip.open
+
+        with xopen(db_name, 'rt') as fp:
+            load_d = json.load(fp)
+            version = load_d['version']
+            assert version == '1.0'
+
+            type = load_d['type']
+            assert type == 'sourmash_lca'
+
+            ksize = int(load_d['ksize'])
+            scaled = int(load_d['scaled'])
+
+            # convert lineage_dict to proper lineages (tuples of LineagePairs)
+            lineage_dict_2 = load_d['lineages']
+            lineage_dict = {}
+            for k, v in lineage_dict_2.items():
+                vv = []
+                for rank in taxlist():
+                    name = v.get(rank, '')
+                    vv.append(LineagePair(rank, name))
+
+                lineage_dict[int(k)] = tuple(vv)
+
+            # convert hashval -> lineage index keys to integers (looks like
+            # JSON doesn't have a 64 bit type so stores them as strings)
```

Comment:
Wait, what? JS doesn't support 64-bit ints, but the python JSON module does the right thing and save the number anyway (it should be up to JS to do the proper thing).

(NOTE: all signatures would be wrong if this assumption was true...)